src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+            assert(forgotExit);
+            exitCount = 2;
+        }
 
-                        case BBJ_THROW:
-                        case BBJ_RETURN:
-                            /* those are exits from the loop */
-                            exit = loopBlock;
-                            exitCount++;
-                            break;
+        // Loop compaction was successful
+        return true;
+    }
 
-                        case BBJ_SWITCH:
+    //------------------------------------------------------------------------
+    // findInsertionPoint: Find an appropriate spot to which blocks that are
+    //    lexically between TOP and BOTTOM but not part of the flow cycle
+    //    can be moved.
+    //
+    // Return Value:
+    //    Block after which to insert moved blocks.
+    //
+    BasicBlock* findInsertionPoint()
+    {
+        // Find an insertion point for blocks we're going to move.  Move them down
+        // out of the loop, and if possible find a spot that won't break up fall-through.
+        BasicBlock* moveAfter = bottom;
+        while (moveAfter->bbFallsThrough())
+        {
+            // Keep looking for a better insertion point if we can.
+            BasicBlock* newMoveAfter = tryAdvanceInsertionPoint(moveAfter);
 
-                            unsigned jumpCnt;
-                            jumpCnt = loopBlock->bbJumpSwt->bbsCount;
-                            BasicBlock** jumpTab;
-                            jumpTab = loopBlock->bbJumpSwt->bbsDstTab;
+            if (newMoveAfter == nullptr)
+            {
+                // Ran out of candidate insertion points, so just split up the fall-through.
+                return moveAfter;
+            }
 
-                            do
-                            {
-                                noway_assert(*jumpTab);
-                                exitPoint = *jumpTab;
+            moveAfter = newMoveAfter;
+        }
 
-                                if (exitPoint->bbNum < top->bbNum || exitPoint->bbNum > bottom->bbNum)
-                                {
-                                    exit = loopBlock;
-                                    exitCount++;
-                                }
-                            } while (++jumpTab, --jumpCnt);
-                            break;
+        return moveAfter;
+    }
 
-                        default:
-                            noway_assert(!"Unexpected bbJumpKind");
-                            break;
-                    }
-                }
+    //------------------------------------------------------------------------
+    // tryAdvanceInsertionPoint: Find the next legal insertion point after
+    //    the given one, if one exists.
+    //
+    // Arguments:
+    //    oldMoveAfter - Prior insertion point; find the next after this.
+    //
+    // Return Value:
+    //    The next block after `oldmoveAfter` that is a legal insertion point
```

Comment:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/29ce6e11b8cf5112b4ffca424e96441dfe7b4f10